### DISCUSSION TITLE
How to enable requests logging?

### DISCUSSION THREAD

**Discussion by msurdi** [vercel/next.js#22945] [Help]
I wonder if when starting a service in production (with `next start`, not with a custom server) there is any way to have in stdout a log of all http requests the server is handling. Even if those requests are serving pre-built static pages and/or SSR pages.

I'm talking about logging the usual:

`info: 127.0.0.1 - - [10/Mar/2021:14:02:25 +0000] "GET /something HTTP/1.1" 404 150 "-" "curl/7.64.1"`

thanks!

**Comment by andrewfritz86** [comment]
IMO this is something that deserves a better explanation in the docs, in that I don't think Next exactly supports this out of the box. 
We have been using middleware to handle this.

```
// middleware.js

import { NextResponse } from 'next/server';

export async function middleware({ nextUrl, cookies }) {
  const response = NextResponse.next();
  console.log(JSON.stringify(nextUrl));
  return response;
}
```

We have been pulling individual keys off the request object and `stringify`ing them, since the whole object doesn't seem to serialize properly.

**Reply by seitbekir** [reply]
What if I need restricted middleware for bunch of urls, but also I need logging.

**Reply by laureatresilience** [reply]
With this option, you also cannot get the status code of the response (or any other information about the response).

**Comment by tomups** [comment]
You can patch the Next.js server to have full access to the request and response objects, and log them with pino-http for example:

https://www.tomups.com/posts/log-nextjs-request-response-as-json/

**Comment by sergiotapia** [comment]
this has to be a joke right, nextjs can't NOT have logging??? I'm coming from elixir and i feel like i'm losing my mind.

**Comment by realtebo** [comment]
Same question here

We are developing a 3rd party integration. So it's fundamental to have full log of method, url with query parameters, headers and body.

**Comment by tybro0103** [comment]
Simply unbelievable.

Honest question - is [OpenTelemetry](https://nextjs.org/docs/app/building-your-application/optimizing/open-telemetry) supposed to be the answer?

**Reply by tomups** [reply]
I could not find a way to log request / response body with OpenTelemetry, so I think for now patching is still the only way if you need that :(

**Comment by ogelami** [comment]
I decided to take a stab at it pull request here https://github.com/vercel/next.js/pull/74438, logging is optional and initiated with `next start --access-log path/file.log`
